test/e2e/clusterctl_upgrade.go

```diff
@@ -299,3 +330,73 @@ func downloadToTmpFile(url string) string {
 
 	return tmpFile.Name()
 }
+
+// deleteAllClustersAndWait deletes all v1alpha3 cluster resources in the given namespace and waits for them to be gone.
+func deleteAllClustersAndWait(ctx context.Context, input framework.DeleteAllClustersAndWaitInput, intervals ...interface{}) {
```

Comment:
Q: should this be implemented as a helper in https://github.com/kubernetes-sigs/cluster-api/blob/master/test/framework/cluster_helpers.go (so we can benefit from `GetAllClustersByNamespace`, `DeleteCluster`, `WaitForClusterDeleted` already existing there)?

Reply:
Can we use those? The new funcs are all using v1alpha3 types which we kind of only need here.

Reply:
@fabriziopandini more than happy to move it to the framework, my initial thought was that we could add the `discovery.ServerSupportsVersion()` check into the main helper functions, that way we could even use most of those helpers in the pre-upgrade steps in the clusterctl_upgrade test, but the existing method signatures complicated that a bit.

I'm more than happy to try and explore either that route or just add parallel methods where needed for the "old" api version. Just let me know which approach you prefer.

Reply:
This is a good point.
What about adding a v1alpa3 suffix to those func so it is more evident why we are using custom helpers here

Reply:
@detiber @fabriziopandini 

Just another idea. What about implementing version-independent deletion helpers? Looks like we only access a few fields during deletion, so I guess it can be reasonably easy implemented with unstructured.

(Feel free to ignore if that would be too hacky :)

Reply:
@sbueringer I'm going to experiment with doing version independence with the framework helpers, but will separate that work out into it's own PR if it works out.

Reply:
Sounds good to me. I didn't see your last answer before sending mine :)